README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
petroOne
========

The goal of petroOne is to ...

Installation
------------

You can install petroOne from github with:

``` r
# install.packages("devtools")
devtools::install_github("f0nzie/petroOne")
```

Examples
--------

### Get the number of papers for *neural networks*.

``` r
library(petroOne)

my_url <- make_search_url(query = "neural network", how = "any")
my_url
#> [1] "https://www.onepetro.org/search?q=neural+network&peer_reviewed=&published_between=&from_year=&to_year="
get_papers_count(my_url)
#> [1] 3314
# 3284
# 3314
```

### Read papers from from\_year to to\_year

We can send a query where we specify the starting years and the end year. Use the parameters as in the example below.

``` r
library(petroOne)

# neural network papers from 1990 to 2000. Exact phrase
my_url <- make_search_url(query = "neural network", 
                          from_year = 1990, 
                          to_year   = 1999, 
                          how = "all")

get_papers_count(my_url)
#> [1] 415
# 415
# 510
onepetro_page_to_dataframe(my_url)
#> # A tibble: 10 x 6
#>                                                      title_data
#>                                                           <chr>
#>  1                          Deconvolution Using Neural Networks
#>  2                     Neural Network Stacking Velocity Picking
#>  3                     Drill-Bit Diagnosis With Neural Networks
#>  4  Seismic Principal Components Analysis Using Neural Networks
#>  5             Neural Networks And Paper Seismic Interpretation
#>  6                    First Break Picking Using Neural Networks
#>  7      Artificial Intelligence I Neural Networks In Geophysics
#>  8         Inversion of Seismic Waveforms Using Neural Networks
#>  9                    Neural Networks In the Petroleum Industry
#> 10 Reservoir Characterization Using Feedforward Neural Networks
#> # ... with 5 more variables: paper_id <chr>, source <chr>, type <chr>,
#> #   year <chr>, author1_data <chr>
```

Get papers by document type (dc\_type)
--------------------------------------

We can also get paper by the type of document. In this example we are requesting only "conference-paper" type.

``` r
# specify document type = "conference-paper", rows = 1000

my_url <- make_search_url(query = "neural network", 
                          how = "all",
                          dc_type = "conference-paper",
                          rows = 1000)

get_papers_count(my_url)
#> [1] 2687
# 2661
# 2687
onepetro_page_to_dataframe(my_url)
#> # A tibble: 1,000 x 6
#>                                                         title_data
#>                                                              <chr>
#>  1                             Deconvolution Using Neural Networks
#>  2                                         Neural Networks And AVO
#>  3                        Neural Network Stacking Velocity Picking
#>  4        Predicting Wax Formation Using Artificial Neural Network
#>  5     Seismic Principal Components Analysis Using Neural Networks
#>  6                Neural Networks And Paper Seismic Interpretation
#>  7        Dynamic Neural Network Calibration of Quartz Transducers
#>  8           Estimation of Welding Distortion Using Neural Network
#>  9 Minimum-variance Deconvolution Using Artificial Neural Networks
#> 10                       First Break Picking Using Neural Networks
#> # ... with 990 more rows, and 5 more variables: paper_id <chr>,
#> #   source <chr>, type <chr>, year <chr>, author1_data <chr>
```

In this other example we are requesting for "journal-paper" type of papers.

``` r
# specify document type = "journal-paper", rows = 1000

my_url <- make_search_url(query = "neural network", 
                          how = "all",
                          dc_type = "journal-paper",
                          rows = 1000)

get_papers_count(my_url)
#> [1] 304
# 303
# 304
onepetro_page_to_dataframe(my_url)
#> # A tibble: 304 x 6
#>                                                                     title_data
#>                                                                          <chr>
#>  1                Artificial Neural Networks Identify Restimulation Candidates
#>  2                                    Drill-Bit Diagnosis With Neural Networks
#>  3                   Implicit Approximation of Neural Network and Applications
#>  4             Application of Artificial Neural Network to Pump Card Diagnosis
#>  5        Application of Artificial Neural Networks to Downhole Fluid Analysis
#>  6           Pseudodensity Log Generation by Use of Artificial Neural Networks
#>  7                Neural Network Approach Predicts U.S. Natural Gas Production
#>  8          An Artificial Neural Network Based Relative Permeability Predictor
#>  9                 Neural Networks for Predictive Control of Drilling Dynamics
#> 10 Characterize Submarine Channel Reservoirs: A Neural- Network-Based Approach
#> # ... with 294 more rows, and 5 more variables: paper_id <chr>,
#> #   source <chr>, type <chr>, year <chr>, author1_data <chr>
```

What to do when we have more than a thousand papers
---------------------------------------------------

### 1. Get total number of papers

First, we need to find what is the total number of papers for a particular query. In this search we want all papers that have exact words "neural network".

``` r

my_url <- make_search_url(query = "neural network", 
                          how = "all")

get_papers_count(my_url)
#> [1] 3025
# 3025 @ 200171009   total number of papers
```

### 2. Find type of papers

In this example we investigate the type of papers available for a given search. In this case, we want exact match of "neural network" words.

``` r
result <- read_onepetro(my_url)
summary_by_doctype(result)
#> # A tibble: 6 x 2
#>               name value
#>              <chr> <dbl>
#> 1 Conference paper  2687
#> 2          General     4
#> 3    Journal paper   304
#> 4            Media     2
#> 5            Other     5
#> 6     Presentation    23
```

We see that most numerous papers are from conferences. We will proceed to gran the paper titles of those.

Summaries
---------

Here is a nother example of summaries. In this case, we want papers that contain the exact words "well test".

``` r
library(petroOne)

my_url <- make_search_url(query = "well test", 
                          how = "all")

result <- read_onepetro(my_url)

summary_by_doctype(result)
#> # A tibble: 7 x 2
#>               name value
#>              <chr> <dbl>
#> 1          Chapter     8
#> 2 Conference paper  9298
#> 3          General   193
#> 4    Journal paper  2529
#> 5            Media     5
#> 6            Other     8
#> 7     Presentation    25
summary_by_publisher(result)
#> # A tibble: 20 x 2
#>                                                     name value
#>                                                    <chr> <dbl>
#>  1                          American Petroleum Institute    42
#>  2                   American Rock Mechanics Association    64
#>  3                                             BHR Group    10
#>  4               Carbon Management Technology Conference     1
#>  5         International Petroleum Technology Conference   364
#>  6              International Society for Rock Mechanics    39
#>  7 International Society of Offshore and Polar Engineers    15
#>  8                                    NACE International    45
#>  9                 National Energy Technology Laboratory     8
#> 10                     Offshore Mediterranean Conference    44
#> 11                        Offshore Technology Conference   506
#> 12                                  Oil Industry Journal    14
#> 13                           Petroleum Society of Canada   396
#> 14                    Pipeline Simulation Interest Group     2
#> 15                  Society of Exploration Geophysicists    75
#> 16                        Society of Petroleum Engineers 10078
#> 17      Society of Petrophysicists and Well-Log Analysts   209
#> 18                      Society of Underwater Technology    12
#> 19        Unconventional Resources Technology Conference    60
#> 20                              World Petroleum Congress    77
summary_by_dates(result)
#> # A tibble: 76 x 2
#>          name value
#>         <chr> <dbl>
#>  1 Since 2017   347
#>  2 Since 2016   913
#>  3 Since 2015  1462
#>  4 Since 2014  2019
#>  5 Since 2013  2530
#>  6 Since 2012  3044
#>  7 Since 2011  3497
#>  8 Since 2010  4023
#>  9 Since 2009  4454
#> 10 Since 2008  4865
#> # ... with 66 more rows
summary_by_publications(result)
#> # A tibble: 575 x 2
#>                                                                 name value
#>                                                                <chr> <dbl>
#>  1           10th North American Conference on Multiphase Technology     1
#>  2                                     10th World Petroleum Congress     1
#>  3                                                11th ISRM Congress     1
#>  4                                     11th World Petroleum Congress     4
#>  5                                                12th ISRM Congress     1
#>  6 12th International Conference on Multiphase Production Technology     2
#>  7                                     12th World Petroleum Congress     3
#>  8                13th ISRM International Congress of Rock Mechanics     1
#>  9 13th International Conference on Multiphase Production Technology     1
#> 10                                     13th World Petroleum Congress     3
#> # ... with 565 more rows
```

In this other example, we want papers that containg the word "well" or "test".

``` r
library(petroOne)

my_url <- make_search_url(query = "well test", 
                          how = "any")

result <- read_onepetro(my_url)

summary_by_doctype(result)
#> # A tibble: 8 x 2
#>               name value
#>              <chr> <dbl>
#> 1          Chapter    60
#> 2 Conference paper 86778
#> 3          General   932
#> 4    Journal paper 15825
#> 5            Media     9
#> 6            Other    21
#> 7     Presentation   265
#> 8         Standard    95
summary_by_publisher(result)
#> # A tibble: 22 x 2
#>                                                     name value
#>                                                    <chr> <dbl>
#>  1                          American Petroleum Institute   676
#>  2                   American Rock Mechanics Association  3851
#>  3                  American Society of Safety Engineers  1098
#>  4                                             BHR Group   221
#>  5               Carbon Management Technology Conference    82
#>  6         International Petroleum Technology Conference  1774
#>  7              International Society for Rock Mechanics  3975
#>  8 International Society of Offshore and Polar Engineers  7408
#>  9                                    NACE International  7708
#> 10                 National Energy Technology Laboratory    21
#> # ... with 12 more rows
summary_by_dates(result)
#> # A tibble: 103 x 2
#>          name value
#>         <chr> <dbl>
#>  1 Since 2017  3542
#>  2 Since 2016  8690
#>  3 Since 2015 14264
#>  4 Since 2014 19543
#>  5 Since 2013 24213
#>  6 Since 2012 28710
#>  7 Since 2011 32778
#>  8 Since 2010 36993
#>  9 Since 2009 40482
#> 10 Since 2008 43857
#> # ... with 93 more rows
summary_by_publications(result)
#> # A tibble: 823 x 2
#>                                                                 name value
#>                                                                <chr> <dbl>
#>  1                                                10th ISRM Congress   109
#>  2           10th North American Conference on Multiphase Technology    20
#>  3                                     10th World Petroleum Congress    82
#>  4                                                11th ISRM Congress   123
#>  5                                     11th World Petroleum Congress    74
#>  6                                                12th ISRM Congress   196
#>  7 12th International Conference on Multiphase Production Technology    31
#>  8                                     12th World Petroleum Congress    54
#>  9                13th ISRM International Congress of Rock Mechanics   228
#> 10 13th International Conference on Multiphase Production Technology    25
#> # ... with 813 more rows
```


NEWS.md
## 20171009 0.1.0.9002
* playing with RSelenium to click over the abstract link to get the text.
* Installing a Docker container with Selenium for firfox-standalone.
* Notebooks to play with Rselenium under ./notebooks/01-url_query folder
* new vignette "thousands_to_dataframe" to bind multiple pages into one dataframe. Not all paper types are homogeneous and they return different columns. Until we find the algorithm, we will retrieve papers by a unique document type.
* paper types do not return complete info in first query. It has to be specified in the type combobox. Problem moved to `test_failing.Rmd`
* new vignette `paper_types.Rmd` to show how to obtain papers by document type. The document types that are homogeneous and return in the first query are: conference papers, jornal papers, general papers, presentation papers and chapter papers. Non-homogeneous: other and media types.

## 20171009 0.1.0.9001
* use tibbles to prevent long printing of dataframes
* use as.tibble in functions onepetro_page_to_dataframe and summary_by_xxx
* reordering chunks in README
* improve explanation of what to do when we have more than 1000 papers to retrieve


## 20171009 0.1.0.9000
* use as.tibble to avoid long printing of dataframes
* show first attempt of splitting paper pages in groups of 1000. Using example of "neural networks" which has 2661 conference-papers. For the time being using one type of papers because other document types have a different number of columns and causes a conflict with the dataframe binding. Working on it.
* build site with pkgdown
* add documentation for datasets
* add tolerance to expect_equal because number of paper keeps growing
* Added a `NEWS.md` file to track changes to the package.





tests/testthat.R
library(testthat)
library(petroOne)

test_check("petroOne")


tests/testthat/test_paper_count.R
library(testthat)

# test the following functions:
    # make_search_url
    # get_papers_count
    # read_onepetro



test_that("get_papers_count matches smartwell papers", {
    my_url <- make_search_url(query = "smartwell")
    expect_equal(get_papers_count(my_url), 70)

})


test_that("get_papers_count matches exact *neural network* papers", {
    my_url <- make_search_url(query = "neural network", how = "all")
    # expect_equal(get_papers_count(my_url), 2998)
    expect_equal(get_papers_count(my_url), 3025, tolerance = 50)
})


# test_that("get_papers_count matches *intelligent completion* papers", {
#     my_url <- make_search_url(query = "intelligent completion")
#     expect_equal(get_papers_count(my_url), 4527)
# })
#
#
# test_that("get_papers_count matches exact *intelligent completion* papers", {
#     my_url <- make_search_url(query = "intelligent completion", how = "all")
#     expect_equal(get_papers_count(my_url), 845)
# })
#
#
# test_that("get_papers_count matches exact *heat transfer* papers", {
#     my_url <- make_search_url(query = "heat transfer", how = "all")
#     expect_equal(get_papers_count(my_url), 7858)
# })
#
# test_that("get_papers_count matches *heat transfer* papers", {
#     my_url <- make_search_url(query = "heat transfer", how = "any")
#     expect_equal(get_papers_count(my_url), 13539)
# })
#
# test_that("get_papers_count matches *heat transfer* papers", {
#     my_url <- make_search_url(query = "heat transfer")
#     expect_equal(get_papers_count(my_url), 13539)
# })
#
#
# test_that("get_papers_count matches *digital oilfield* papers", {
#     my_url <- make_search_url(query = "digital oilfield", how = "all")
#     expect_equal(get_papers_count(my_url), 527)
#     # print(get_papers_count(my_url))
# })


tests/testthat/test_document_type.R
library(testthat)

test_that("document type = *journal-paper* matches", {
    my_url <- make_search_url(query = "neural network",
                              how = "all",
                              dc_type = "journal-paper",
                              rows = 1000)

    expect_equal(get_papers_count(my_url), 303, tolerance = 3)
})


test_that("document type = *conference-paper* matches", {
    my_url <- make_search_url(query = "neural network",
                              how = "all",
                              dc_type = "conference-paper",
                              rows = 1000)

    # expect_equal(get_papers_count(my_url), 2661, tolerance = 30)
    expect_equal(get_papers_count(my_url), 2687, tolerance = 30)
})


R/summary.R
library(rvest)


#' @importFrom magrittr %>%
publication_result_right <- function(result) {
    pub_doc <- result %>%
        html_nodes(".facet-unit-right") %>%
        html_nodes("option") %>%
        html_text()
    pub_doc
}


#' @importFrom magrittr %>%
publication_result_left <- function(result) {
    pub_doc <- result %>%
        html_nodes(".facet-unit-left") %>%
        html_nodes("option") %>%
        html_text()
    pub_doc
}

#' @title Get summary dataframe by publisher
#' @description Summary dataframe by publisher: OTC, SPE, etc.
#' @param result a OnePetro page with results
#' @export
summary_by_publisher <- function(result) {
    x <- publication_result_right(result)
    pub_vector <- get_dc_publisher(x)
    value <- extract_num_papers(pub_vector)
    name   <- extract_publishers(pub_vector)
    tibble::as.tibble(data.frame(name, value, stringsAsFactors = FALSE))
}

#' @title Get summary dataframe by document type
#' @description Document types are: conference-paper, journal-paper, etc.
#' @param result a OnePetro page with results
#' @export
summary_by_doctype <- function(result) {
    x <- publication_result_right(result)
    doctype_vector <- get_dctype(x)
    value <- extract_num_papers(doctype_vector)
    name   <- extract_publishers(doctype_vector)
    tibble::as.tibble(data.frame(name, value, stringsAsFactors = FALSE))
}

#' @title Get summary dataframe by year
#' @description Year that the paper was published
#' @param result a OnePetro page with results
#' @export
summary_by_dates <- function(result) {
    x <- publication_result_left(result)
    pub_vector <- get_dc_issued_year(x)
    value <- extract_num_papers(pub_vector)
    name   <- extract_publishers(pub_vector)
    tibble::as.tibble(data.frame(name, value, stringsAsFactors = FALSE))
}

#' @title Get summary dataframe by publication
#' @description Publications could be World Petroleum Congress, Annual Technical
#' Meeting, SPE Unconventional Reservoirs Conference, etc.
#' @param result a OnePetro page with results
#' @export
summary_by_publications <- function(result) {
    x <- publication_result_left(result)
    pub_vector <- get_s2_parent_title(x)
    value <- extract_num_papers(pub_vector)
    name   <- extract_publishers(pub_vector)
    tibble::as.tibble(data.frame(name, value, stringsAsFactors = FALSE))
}


get_dctype <- function(aList) {
    len_list <- length(aList)
    ix <- grep("All types", aList)
    aList[(ix+1):len_list]
}

get_dc_publisher <- function(aList) {
    ix_stop <- grep("All types", aList)
    aList[2:(ix_stop-1)]
}

get_dc_issued_year <- function(aList) {
    ix_stop <- grep("All publications", aList)
    aList[2:(ix_stop-1)]
}

get_s2_parent_title <- function(aList) {
    len_list <- length(aList)
    ix <- grep("All publications", aList)
    aList[(ix+1):len_list]
}



extract_num_papers <- function(x) {
    pattern <- "(?<=\\{).+(?=\\})"
    m <- regexpr(pattern, x, perl = TRUE)
    as.numeric(gsub(",", "" , regmatches(x, m)))
}


extract_publishers <- function(x) {
    # pattern <- "(\\s[{\\d}].+)"
    pattern <- "\\s{([0-9].*)}"
    gsub(pattern, "", x, perl = TRUE)
}





R/petroOne-package.r
#' petroOne.
#'
#' @name petroOne
#' @docType package
NULL


#' @title Discipline and Subject  labels dataset
#' @description dataset containing disciplines and subjects.
#' The purpose is to categorize papers based on the words in their title since
#' OnePetro does not supply keywords or any sort of categorization.
#' File:  disciplines.rda
#' Class: data.frame
#' @docType data
"labels"

#' @title Default custom stop words
#' @description This is a minimal dataset of custom stopwords.
#' Tou can supply your own stopwords by editing the file stopwords.txt and
#' then importing them. But this is a basic way to start and eliminate
#' common words from the paper titles during clasification.
#' Dataset: stopwords.rda
#' Source:  stopwords.txt
#' Class:   char vector
#' @docType data
"custom_stopwords"

#' @title Default synonyms
#' @description Synonyms dataset to prevent repetition of equivalent words
#' during classification. Example: 2D/2-D, cased hole/cased-hole,
#' deep water/deepwater, etc.
#' The first column is for the original word; the second column is for the
#' replacement or standard word.
#' Dataset: synonyms.rda
#' Source:  synonyms.txt
#' Class:   data.frame
#' @docType data
"custom_synonyms"


R/papers_to_dataframe.R

#' @title Reads a OnePetro URL and converts it to a dataframe
#' @description A OnePetro URL with a query is read into a HTML page and
#' converted to a dataframe
#' @param url char a OnePetro type URL
#' @export
onepetro_page_to_dataframe <- function(url) {
    webpage <- read_html(url)
    df_titles  <- read_titles(webpage)
    df_sources <- read_sources(webpage)
    df_author  <- read_author(webpage)

    # ensure that all dataframe have the same number of rows
    if (all(dim(df_titles)[1]  == dim(df_sources)[1],
            dim(df_sources)[1] == dim(df_author)[1],
            dim(df_author)[1]  == dim(df_titles)[1]
    ))
        df <- cbind(df_titles, df_sources, df_author)
    else
        stop("Dataframe sizes different")  # otherwise, stop

    return(tibble::as.tibble(df))
}





onepetro_allpages_to_dataframe <- function(url) {
    # webpage <- read_html(url)
    papers_count <- get_papers_count(url)
    if (papers_count > 1000) {
        num_pages <- papers_count / 1000
    } else {
        num_pages = 1
    }

    info <- list(papers = papers_count, pages1000 = num_pages)

    for (page in seq_len(num_pages)) {
        # webpage <- read_html(url)

    }
    info
}


#' @importFrom rvest html_nodes html_text
read_titles <- function(webpage) {

    # title
    # .result-link

    #Using CSS selectors to scrap the rankings section
    title_data_html <- html_nodes(webpage, '.result-link')

    # Converting the ranking data to text
    title_data <- html_text(title_data_html)

    # data pre-processing
    title_data <- trimws(gsub("\n", "",title_data))


    #Let's have a look at the rankings
    title_data <- data.frame(title_data, stringsAsFactors = FALSE)
    # write.csv(df, file = "3000-conf.csv")
    return(title_data)
}

#' @importFrom rvest html_nodes html_text
read_sources <- function(webpage) {
    # year, paper id, institution, type, year
    # .result-item-source

    #Using CSS selectors to scrap the rankings section
    source_data_html <- html_nodes(webpage, '.result-item-source')

    #Converting the ranking data to text
    source_data <- html_text(source_data_html)
    # print(source_data);    stop()

    # pre-processing. split at \n
    source_data <- data.frame(do.call('rbind', strsplit(as.character(source_data),
                                                        '\n',fixed=TRUE)),
                              stringsAsFactors = FALSE)
    # remove blank columns

    source_data <- source_data[, 2:5]
    # rename columns
    names(source_data) <- c("paper_id", "source", "type", "year")
    # remove dash from year
    source_data$year <- gsub("-", "", source_data$year)

    # Let's have a look at the paper source data
    source_data
}

#' @importFrom rvest html_nodes html_text
read_author <- function(webpage) {
    # author #1
    #Using CSS selectors to scrap the rankings section
    author1_data_html <- html_nodes(webpage, '.result-item-author:nth-child(1)')

    #Converting the ranking data to text
    author1_data <- html_text(author1_data_html)

    # data pre-processing
    author1_data <- trimws(gsub("\n", "", author1_data))

    #Let's have a look at the rankings
    data.frame(author1_data, stringsAsFactors = FALSE)
}


R/url.R

#Loading the rvest package
library(rvest)

# Specifying the url for desired website to be scrapped
url <- "https://www.onepetro.org/search?start=0&q=neural+networks&from_year=&peer_reviewed=&published_between=&rows=999&to_year="

form_input <- list(dummy = "dummy", query = "?q=", peer_reviewed = "peer_reviewed=",
                   published_between = "published_between=",
                   from_year = "from_year=",
                   to_year = "to_year=",
                   start = "start=",
                   rows = "rows=",
                   dc_type = "dc_type=")

# Examples:
# https://www.onepetro.org/search?q=%22data+science%22&peer_reviewed=&published_between=&from_year=&to_year=
#




#' @title Read OnePetro web page given a query URL
#' @description Read a OnePetro webpage using a query URL. Uses xml2
#' function read_html
#' @param url char a query URL for OnePetro
#' @rdname read_onepetro
#' @importFrom xml2 read_html
#' @export
read_onepetro <- function(url) {
    #Reading the HTML code from the website
    read_html(url)
}


#' @title Number of paper for a given query
#' @description Obtains the number of papers being queried by the URL
#' @param url char a query URL for OnePetro
#' @rdname get_papers_count
#' @importFrom magrittr %>%
#' @importFrom rvest html_nodes html_text
#' @export
get_papers_count <- function(url) {
  # result <- send_url(url)
  result <- read_onepetro(url)

  papers <- result %>%
    html_nodes("h2") %>%
    html_text()

  # extract the numeric part of the results
  pattern <- "[\\d,]+(?= results.)"    # a number, including comma, before " results."
  m <- regexpr(pattern, papers[1], perl = TRUE)       # matched
  as.numeric(gsub(",", "", regmatches(papers[1], m))) # remove comma first
}




#' @param query char         any words that will be searched
#' @param start int          optional to set the starting papaper
#' @param from_year int      optional to indicate starting year
#' @param peer_reviewed     bool optional, TRUE or FALSE
#' @param published_between bool automatic if from_year or to_year are on
#' @param rows int           optional. number of papers to retrieve. max=1000
#' @param to_year int        optional to inndicate end year
#' @param dc_type char       optional to indicate if journal, conference paper
#' @param how char           default="any". "all" will match exact words
#' @title Make a search URL for OnePetro
#' @description Create a URL that works in OnePetro website
#' @export
make_search_url <- function(query = NULL, start = NULL, from_year = NULL,
                            peer_reviewed = NULL,
                            published_between = NULL,
                            rows = NULL,
                            to_year = NULL,
                            dc_type = NULL,
                            how = "any") {

    website <- "https://www.onepetro.org"

    if (!is.null(start) || !is.null(rows)) {
        if (!is.null(rows) & is.null(start)) start = 0
        stopifnot(is.numeric(start), is.numeric(rows))
    }

    if (!is.null(from_year) && !is.null(to_year)) {
        stopifnot(is.numeric(from_year), is.numeric(to_year))
    }

    if (is.null(query)) {
        stop("search words not provided")
    } else {
        split_query <- unlist(strsplit(query, " "))
        if (length(split_query) > 1) {
            query <- paste(split_query, collapse = "+")
            # use function shQuote to add extra quotes when we want how = "all"
            #query <- ifelse(how == "all", dQuote(query), query)
            query <- ifelse(how == "all", shQuote(query), query)
            # query <- ifelse(how == "all", paste0("'", query, "'"), query)

            # print(query)
        }
    }

    if (!is.null(from_year) || !is.null(to_year)) {
        # use regex to validate year is between 1900 and 2099
        pattern <- "(?:(?:19|20)[0-9]{2})"
        if (!grepl(pattern, from_year, perl = TRUE) ||
            !grepl(pattern, to_year,   perl = TRUE)) stop("year not valid")
        # if valid year then turn on published_between
        published_between = "on"
        # if any of the *from* or *to* years are null replace with empty char
        if (is.null(from_year)) {
            from_year = ""
        }
        if (is.null(to_year)) {
            to_year = ""
        }
    }

    # peer_reviewed=on if TRUE; blank if unslected or FALSE
    if (is.null(peer_reviewed)) {
        peer_reviewed = ""
    } else {
        if (peer_reviewed) peer_reviewed = "on"
    }

    # document type
    if (!is.null(dc_type)) {
        valid_options <- c("conference-paper", "journal-paper", "general",
                           "media", "general", "presentation", "chapter", "other")
        # stop if it is not in the options
        if (!dc_type %in% valid_options) {
            msg <- sprintf("Option unknown. It must be one of [ %s ]",
                           paste(valid_options, collapse = ", "))
            stop(msg)
            # cat(valid_options, "\n")
        }
    }

    s_search  <- paste(website, "search", sep = "/")

    # these strings will need to join with the ampersand & at the tail
    s_query   <- paste0("?q=", query)
    s_peer    <- paste0("peer_reviewed=", peer_reviewed)
    s_publish <- paste0("published_between=", published_between)
    s_from    <- paste0("from_year=", from_year)
    s_to      <- paste0("to_year=", to_year)
    s_start   <- paste0("start=", start)
    s_rows    <- paste0("rows=", rows)
    s_type    <- paste0("dc_type=", dc_type)

    # url
    s_url <- list(websearch = s_search, query = s_query, peer = s_peer,
                  published_between = s_publish, from_year = s_from, to_year = s_to,
                  start = s_start, rows = s_rows, dc_type = s_type
    )

    for (i in 1:length(s_url)) {
        # cat(i, my_url[[i]], "\n")
        if (i == 1) joined <- s_url[[i]]
        if (i == 2) joined <- paste0(joined, s_url[[i]])
        if (i >=3 ) {
            if (s_url[[i]] == form_input[[i]] & i <= 6) {
                joined <- paste(joined, s_url[[i]], sep = "&")
            } else  if (s_url[[i]] != form_input[[i]]) {
                # cat(i, s_url[[i]], "\n")
                joined <- paste(joined, s_url[[i]], sep = "&")
            }
        }
    }

    joined
}








create_url <- function(start = NULL, query = NULL, from_year = NULL,
                       peer_reviewed = NULL,
                       published_between = NULL,
                       rows = NULL,
                       to_year = NULL,
                       how = "any") {

    website <- "https://www.onepetro.org"

    if (is.null(start)) {
        start = ""
    }
    if (is.null(query)) {
        stop("search words not provided")
    } else {
        split_query <- unlist(strsplit(query, " "))
        if (length(split_query) > 1) {
            query <- paste(split_query, collapse = "+")
            # use function shQuote to add extra quotes when we want how = "all"
            query <- ifelse(how == "all", shQuote(query), query)
            print(query)
        }
    }
    print(query)

    if (is.null(from_year)) {
        from_year = ""
    }
    if (is.null(peer_reviewed)) {
        peer_reviewed = ""
    }
    if (is.null(published_between)) {
        published_between = ""
    }
    if (is.null(rows)) {
        rows = ""
    } else {
        if(is.null(start)) start = 0
    }
    if (is.null(to_year)) {
        to_year = ""
    }

    s_search  <- paste(website, "search", sep = "/")
    s_q       <- paste0("?q=", query)
    s_peer    <- paste0("peer_reviewed=", peer_reviewed)
    s_publish <- paste0("published_between=", published_between)
    s_from    <- paste0("from_year=", from_year)
    s_to      <- paste0("to_year=", to_year)
    s_start   <- paste0("start=", start)
    s_rows    <- paste0("rows=", rows)

    url <- paste(s_q, s_peer, s_publish, s_from, s_to, sep = "&")
    url <- paste0(s_search, url)
    url
}


send_url <- function(url, how = "any") {
    #Reading the HTML code from the website
    read_html(url)
}
